Clear up import policy

Bibo-Joshi opened the pull request:
closes #2468

Some notes:

- I know 120 files sounds bad the the majority of changes is really just imports :D 
- `ext.utils.promise.Promise` was deliberately not moved to `ext.promise.Promise` because since the removal of `MessageQueue` users should almost never have contact with that class. You can view that at [RTD](https://python-telegram-bot.readthedocs.io/en/clear-up-imports/telegram.html#auxiliary-modules)
- While at it I split up that messy `tg.utils.helpers` module into a number of separate modules and moved parts of the contents just to the place where it's actually used. Also added a bunch of additional tests
- Added a new section in the docs for the modules in `telegram` that are not strictly API related

TODO:

- [x] check if we can add some more `versionchanged` directives
- [x] Wait for #2662 to me merged so that we can include the changes to the warnings
- [x] There's now `tg.helpers` and `tg.utils.helpers` maybe we can find a better name for one of the two to avoid confusion …

### Checklist for PRs

- [x] Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)  -- This can probably still be improved

harshil21 reviewed (changes requested):
Good changes. Some things we can still change:

- While at it, we can also rename base.py to telegramobject.py, cause for some reason in rtd the 'bases' section doesn't link is broken (just updated my sphinx to 4.2.0 from 4.0.2 and now this is fixed!), and I suppose telegramobject.py is more explicit.

- Also `telegram.ext.filters Module` is listed under the ['Handlers' section](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.html#handlers). Maybe move it up where the rest of telegram.ext is (i.e with Updater and the rest)?

- I noticed `telegramobject.rst` doesn't have `:github_url:`, maybe add that?

Bibo-Joshi commented on telegram/ext/commandhandler.py:
```diff
@@ -53,7 +53,7 @@ class CommandHandler(Handler[Update, CCT]):
 
     Args:
-        command (:class:`telegram.utils.types.SLT[str]`):
+        command (:obj:`str` | Tuple[:obj:`str`] | List[:obj:`str`]):
```
because of that and because `utils.types` is not part of the public API. don't know why I agreed to adding it in the first place :D

Bibo-Joshi commented on telegram/ext/updater.py:
```diff
@@ -38,19 +38,29 @@
     from telegram.ext import BasePersistence, Defaults, CallbackContext
 
 
+# From https://stackoverflow.com/questions/2549939/get-signal-names-from-numbers-in-python
```
I was unsure, because that's not documented and because of the comment about pylint. But it seems that it's indeed just not documented, but it does exist:

* https://github.com/python/typeshed/pull/555#issuecomment-247874222
* https://bugs.python.org/issue28206

There was a [PR](https://github.com/python/cpython/pull/1939) which never made it for some reason - maybe I'll pick it up 😄 

TL;DR: will change

Bibo-Joshi commented on telegram/message.py:
```diff
@@ -17,7 +17,13 @@
-"""This module contains an object that represents a Telegram Message."""
+"""This module contains an object that represents a Telegram Message.
+
+Warning:
+    Contents of this module are intended to be used internally by the library and *not* by the
+    user. Changes to this module are not considered breaking changes and may not be documented in
+    the changelog.
```
![image](https://user-images.githubusercontent.com/22366557/134295834-a50bac32-9efb-48ea-b07e-4bd7351723d9.png)

Me trying to figure out why I did that

Bibo-Joshi commented:
> * While at it, we can also rename base.py to telegramobject.py, cause for some reason in rtd the 'bases' section doesn't link is broken (just updated my sphinx to 4.2.0 from 4.0.2 and now this is fixed!), and I suppose telegramobject.py is more explicit.

The broken link is not really related to the file name, but I renamed anyway. see sphinx-doc/sphinx#9026
But this is actually a good reminder to follow up on sphinx-doc/sphinx#4944 being fixed now 👍 we should do that in the doc-fixes branch though

> * Also `telegram.ext.filters Module` is listed under the ['Handlers' section](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.html#handlers). Maybe move it up where the rest of telegram.ext is (i.e with Updater and the rest)?

I moved it there on purpose a while back, because `filters` basically belongs to `MessageHandler` … at least for me that makes more sense …

> * I noticed `telegramobject.rst` doesn't have `:github_url:`, maybe add that?

✔️

harshil21 commented on telegram/utils/warnings.py:
```diff
@@ -61,8 +38,10 @@ def warn(message: str, category: Type[Warning] = PTBUserWarning, stacklevel: int
-            same value as you'd pass directly to ``warnings.warn()``.
+        message (:obj:`str`): Specify the warnings message to pass to ``warnings.warn()``.
+        category (:obj:`Type[Warning]`, optional): Specify the Warning class to pass to
+            ``warnings.warn()``. Defaults to :class:`telegram.warnings.PTBUserWarning``
```
```suggestion
            ``warnings.warn()``. Defaults to :class:`telegram.warnings.PTBUserWarning`
```
